man/rmsgw_aci.1
.\"
.\"			r m s g w _ a c i . 1
.\" $Revision: 67 $
.\" $Author: eckertb $
.\" $Id: rmsgw_aci.1 67 2008-08-21 18:04:19Z eckertb $
.\"
.\" RMS Gateway
.\"
.\" Copyright (c) 2004-2008 Hans-J. Barthen - DL5DI
.\" Copyright (c) 2008 Brian R. Eckert - W3SG
.\"
.\" Questions or problems regarding this program can be emailed
.\" to linux-rmsgw@w3sg.org
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH RMSGW_ACI 1 "Linux RMS Gateway" "W3SG/DL5DI" "W3SG/DL5DI"
.SH NAME
rmsgw_aci \- Linux RMS Gateway Automatic Check-In Program
.SH SYNOPSIS
.B rmsgw_aci -h
.br
.B rmsgw
.RB [ \-c
.IR configfile ]
.RB [ \-l
.IR logmask ]
.SH DESCRIPTION
The 
.B rmsgw_aci
program performs an automatic check-in to the Winlink2000 system for
the Linux RMS Gateway to update the status of all defined channels (see
.IR rms-channels (5)).
.P
All logging for the program (and gateway as a whole) is done
through the standard
.BR syslog (3)
facility, and must be configured along with the gateway itself for
proper logging operation.
Selection of the syslog facility is done through the gateway's
configuration file (see
.IR rms-config (5)),
and the priority logmask is selectable either through the config file or
the
.B \-l
command line option.
.P
The 
.B rmsgw_aci
program is run periodically through cron,
and is best run every 30 minutes,
but on some arbitrary minute of the hour, rather than
exactly on the hour and half-hour. If everyone configured their automatic
check-in to run exactly on the same hour and half-hour,
the CMS's would take a large "hit" all at once, which could
have potential performance issues for the Winlink2000 system.
.P
Your choice of the start for the 30 minute interval to run the automatic
check-in is entirely up to you.
.P
The CMS host list used for the check-in is specified in the
.I /etc/rmsgw/hosts
file (see
.BR cms-hosts (5)),
which contains CMS port and password data along with the CMS host name.
The check-in, along with normal gateway activity,
will round-robin its connections among the configured hosts.
The status files for each host are maintained in
.IR /etc/rmsgw/stat ,
where a file entry for each CMS host name will appear if the host has
had a connection or attempted connection. The access time of the file
indicates when the host was last tried,
and the automatic check-in will attempt to connect to the CMS host that was
least recently tried.
.P
The cron entry should be created for the user that is created during the
gateway installation and setup
.RI ( rmsgw
by default).
.SH EXAMPLE CRONTAB ENTRY
.nf
12,42 * * * * /usr/local/bin/rmsgw_aci >/dev/null 2>&1
.fi
.SH OPTIONS
.TP
.BI \-c\  configfile
Instead of using the standard configfile,
.IR /etc/rmsgw/gateway.conf ,
the file indicated by
.I configfile
is used to set the run-time configuration.
.TP
.BI \-l\  logmask
Set the
.BR syslog (3)
logmask for the automatic check-in. Setting the logmask determines the lowest
.BR syslog (3)
.I priority
that will be written to the logfile. The recognized strings,
in order of their priority from highest to lowest are:
.BR alert ,
.BR crit ,
.BR emerg ,
.BR err ,
.BR warning ,
.BR notice ,
.BR info ,
and
.BR debug .
As you lower the priority, more and more messages will be logged via syslog.
The
.B debug
priority is useful when trying to determine the source of problems either
during development and testing, or during the initial setup and testing;
it should not be used for normal operations. The default is, and most sites
will likely want to use
.BR info .
.IP
The basic check-in messages are logged with a priority of
.BR notice .
Setting the
.B \-l
option on the command line over-rides the internal default and the
.B LOGMASK
setting in
.IR /etc/rmsgw/gateway.conf ,
or a config file specified using
.BR \-c .
See
.BR rms-config (5).
.SH FILES
.TP
.I /etc/rmsgw/gateway.conf
The default gateway configuration file.
See
.BR rms-config (5).
.TP
.I /etc/rmsgw/hosts
The CMS host table.
See
.BR cms-hosts (5).
.TP
.I /etc/rmsgw/channels.xml
The gateway channels configuration xml document.
See
.BR rms-channels (5).
.TP
.I /etc/rmsgw/acihelp
General help file displayed when executing
.BR \"rmsgw -h\" .
.SH SEE ALSO
.BR rmsgw (1),
.BR syslog (3),
.BR syslog.conf (5),
.BR syslogd.conf (5),
.BR ax25d.confg (5),
.BR rms-config (5),
.BR cms-hosts (5),
.BR rms-channels (5)
.SH AUTHORS
Hans-J. Barthen - DL5DI <dl5di@gmx.de>
.br
Brian R. Eckert - W3SG <eckertb@w3sg.org>



rmsgw_aci/rmsgw_aci.c
/*
 *			r m s g w _ a c i . c
 * $Revision: 176 $
 * $Author: eckertb $
 *
 * RMS Gateway
 *
 * Copyright (c) 2004-2009 Hans-J. Barthen - DL5DI
 * Copyright (c) 2008-2009 Brian R. Eckert - W3SG
 *
 * Questions or problems regarding this program can be emailed
 * to linux-rmsgw@w3sg.org
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#ifndef lint
static char svnid[] = "$Id: rmsgw_aci.c 176 2014-10-27 09:07:54Z eckertb $";
#endif /* lint */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <syslog.h>
#include <errno.h>
#include <limits.h>

#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>

#include "rms.h"
#include "rmslib.h"
#include "mapname.h"

#include "aci.h"

/*
 * aci globals
 */
char *active_configfile = NULL; /* the config file actually being used */

/*
 * externs
 */
extern int shm_debug;


/***
 *  dohelp()
 */
static void dohelp(char *helpfile)
{
     if (printfile(helpfile) < 0) {
	  fprintf(stderr, "Can't dislay helpfile '%s' (errno = %d)\n",
		  helpfile, errno);
     }
}


/***
 *  main()
 */
int main(int argc, char *argv[])
{
     char *gwcallarg = NULL; /* command line supplied gateway callsign */
     char *logmaskarg = NULL; /* command line over-ride of logmask */
     char *configfilearg = NULL; /* command line over-ride of config file */
     config *cfg;
     time_t age;
     char statfile[PATH_MAX];
     int rc; /* generic return code var */
     int opt;
     version_blk *version;
     time_t version_update_interval = (time_t) 0;
     extern int optind;
     extern char *optarg;
     extern rms_status *rms_stat;

     /*
      * set shared memory activity debugging flag
      */
     shm_debug = set_shm_debug();

     /*
      * intially open the log and set the default log mask
      * to get started and until we get the configuration
      */
     openlog("rmsgw_aci", LOG_CONS|LOG_PID|LOG_NDELAY, LOG_LOCAL0);
     setlogmask(LOG_UPTO(LOG_INFO));

     /*
      * parse the command line options and take appropriate actions
      */
     while ((opt = getopt(argc, argv, "hc:l:g:")) != -1) {
	  switch (opt) {
	  case 'h': /* help */
	       dohelp(ACIHELPFILE);
	       exit(0);
	       break;
	  case 'c': /* config file */
	       configfilearg = strdup(optarg);
	       break;
	  case 'l': /* log mask */
	       logmaskarg = strdup(optarg);
	       break;
	  case 'g': /* gateway call */
	       gwcallarg = strdup(optarg);
	       break;
	  case ':':
	  case '?':
	  default:
	       syslog(LOG_ERR, "Invalid option usage");
	       exit(1);
	  }
     }

     /*
      * get the version info
      */
     if ((version = loadVersion(RMSVERSIONFILE, &rc)) == NULL) {
	  switch (rc) {
	  case ERR_VER_MISSINGPACK:
	       syslog(LOG_ERR, "ERROR: Version: Missing Package Name!");
	       break;
	  case ERR_VER_MISSINGPROG:
	       syslog(LOG_ERR, "ERROR: Version: Missing Program Name!");
	       break;
	  case ERR_VER_MISSINGLAB:
	       syslog(LOG_ERR, "ERROR Version: Missing Version Label!");
	       break;
	  case ERR_VER_MAPERROR:
	       syslog(LOG_ERR,
		      "ERROR: in version file: %s", maperrmsg(map_errno));
	       break;
	  default: /* should never happen! */
	       syslog(LOG_CRIT, "UNKNOWN ERROR (%d) IN VERSION FILE", rc);
	       break;
	  }

	  fprintf(stderr, "%s: failed to load version info. See syslog.\n",
		  argv[0]);
	  exit(1);
     }

     /*
      * get the configuration
      */
     active_configfile = configfilearg ? configfilearg : configfile;
     if ((cfg = loadConfig(active_configfile, &rc)) == NULL) {
	  switch (rc) {
	  case ERR_CFG_MISSINGFILE:
	       syslog(LOG_ERR, "ERROR: missing config file");
	       break;
	  case ERR_CFG_INVALIDCALL:
	       syslog(LOG_ERR, "ERROR: invalid gateway callsign");
	       break;
	  case ERR_CFG_MISSINGCALL:
	       syslog(LOG_ERR,
		      "ERROR: missing gateway callsign in config");
	       break;
	  case ERR_CFG_MAPERROR:
	       syslog(LOG_ERR,
		      "ERROR: config file: %s", maperrmsg(map_errno));
	       break;
	  default: /* should never happen! */
	       syslog(LOG_CRIT, "UNKNOWN ERROR IN CONFIG FILE");
	       break;
	  }
	  fprintf(stderr, "%s: failed to load configuration. See syslog.\n",
		  argv[0]);
	  exit(1);
     }

     /*
      * replace configured values with any supplied command
      * line over-rides
      */
     if (gwcallarg) { /* gateway call over-ride? */
	  free(cfg->gwcall);
	  cfg->gwcall = gwcallarg;
     }

     if (logmaskarg) { /* logmask over-ride? */
	  free(cfg->logmask);
	  cfg->logmask = logmaskarg;
     }

     /*
      * reopen syslog using the configured facility/logmask
      */
     closelog();
     openlog("rmsgw_aci", LOG_CONS|LOG_PID|LOG_NDELAY,
	     mapfacility(downcase(cfg->logfacility)));
     setlogmask(LOG_UPTO(mappriority(downcase(cfg->logmask))));

     syslog(LOG_INFO,"%s - %s ACI %s %s (%s)",
 	    cfg->gwcall, version->package, version->label,
	    __DATE__, cfg->gridsquare);

     /*
      * check for unconfigured or improperly configured gateway
      * (this check is rather simplistic, but will catch a basic
      *  case of an incomplete configuration)
      */
     if (!strcmp(cfg->gwcall, "N0CALL")) {
	  syslog(LOG_CRIT,
		 "FATAL: Incomplete/incorrect gateway configuration (N0CALL)");
	  exit(1);
     }

     /*
      * attach the status block to the process (attached memory address will
      * be at returned value in rms_stat)
      */
     if (statblock_attach(&rms_stat) < 0) {
	  syslog(LOG_CRIT, "FATAL: cannot attach shared memory segment");

	  fprintf(stderr,
		  "FATAL error: can't attach status block (see syslog)\n");

	  closelog();
	  exit(1);
     }

     /*
      * update status block - set initial condition
      */
     sbset_aci_pid(rms_stat);
     sbset_aci_state(rms_stat, ACI_RUNNING);

     /*
      * set any environment variables for this execution environment
      */
     if (cfg->rmsgwacienv != NULL) {
	  if (file_exists(cfg->rmsgwacienv)) {
	       if (loadenv(cfg->rmsgwacienv) != SUCCEED) {
		    syslog(LOG_WARNING, "unable to set environment variables");
	       }
	  }
     }
	 
     /*
      * call gateway startup hook
      */
     runhook(cfg->hookdir, START_RMSGW_ACI, active_configfile, NULL);

     /*
      * update channel status
      */
     if (wl2k_aci(cfg, rms_stat) < 0) {
	  sbset_aci_state(rms_stat, ACI_IDLE);
	  statblock_detach(rms_stat);
	  exit(1); /* failed aci */
     }/* else... */

     /*
      * get the update interval for version info
      */
     if (cfg->versionupd == NULL || ((version_update_interval = atol(cfg->versionupd)) == 0)) {
	  version_update_interval = DFLT_VERSIONUPD;
     }
     syslog(LOG_DEBUG, "Version update interval = %ld", version_update_interval);
       
     /*
      * check status file for version update -
      * if the version hasn't been updated in an day or more,
      * send it
      */
     sprintf(statfile, "%s/.version.%s", CMSSTATDIR, cfg->gwcall);
     age = fileage(statfile);
     syslog(LOG_DEBUG, "Version file '%s' age %ld", statfile, age);

     if (age >= version_update_interval) {
	  sbset_aci_state(rms_stat, ACI_VERSIONUPDATE);

	  runhook(cfg->hookdir, PRE_RMSGW_ACI_VERSION_UPDATE, active_configfile, statfile, NULL);

	  sendVersion(cfg);
	  sbset_aci_state(rms_stat, ACI_RUNNING);

	  /*
	   * update the status file
	   */
	  syslog(LOG_DEBUG, "touch %s", statfile);
	  setstatfile(statfile);

	  runhook(cfg->hookdir, POST_RMSGW_ACI_VERSION_UPDATE, active_configfile, statfile, NULL);
     }

     sbset_aci_state(rms_stat, ACI_IDLE);
     statblock_detach(rms_stat);

     runhook(cfg->hookdir, END_RMSGW_ACI, active_configfile, NULL);

     exit(0);
}
